runtime/lua/vim/keymap.lua

```diff
@@ -0,0 +1,223 @@
+local keymap = {}
+
+-- Have to use a global to handle re-requiring this file and losing all of the keymap.
+--  In the future, the C code will handle this.
+__VimMapStore = __VimMapStore or {}
+keymap._store = __VimMapStore
+
+keymap._create = function(f)
+  table.insert(keymap._store, f)
+  return #keymap._store
+end
+
+keymap._execute = function(id)
+  keymap._store[id]()
+end
+
+local make_mapper = function(mode, defaults, opts)
+  local args, map_args = {}, {}
+  for k, v in pairs(opts) do
+    if type(k) == 'number' then
+      args[k] = v
+    else
+      map_args[k] = v
+    end
+  end
+
+  local lhs = opts.lhs or args[1]
+  local rhs = opts.rhs or args[2]
+  local map_opts = vim.tbl_extend("force", defaults, map_args)
+
+  local mapping
+  if type(rhs) == 'string' then
+    mapping = rhs
+  elseif type(rhs) == 'function' then
+    assert(map_opts.noremap, "If `rhs` is a function, `opts.noremap` must be true")
+
+    local func_id = keymap._create(rhs)
+    mapping = string.format(
+      [[<cmd>lua vim.keymap._execute(%s)<CR>]], func_id
```

Comment:
This mapping will fail if `expr = true` in the opts.

Reply:
Hmm nevermind it looks like there are some quirks with the way mappings are handled so this

```lua
_G.fn = function()
  return '<tab>'
end
vim.api.nvim_set_keymap('i', '<tab>', 'luaeval("fn()")', {expr=true})
```

will insert the string `<tab>` whereas this

```lua
vim.api.nvim_set_keymap('i', '<tab>', 'luaeval("string.lower(\'<tab>\')")', {expr=true})
```

inserts an actual tab.